File: test/javascript/deserializer_test.mjs
"use strict";

import {
  assert,
  contextFixture,
  defineGlobalErlangAndElixirModules,
} from "./support/helpers.mjs";

import Deserializer from "../../assets/js/deserializer.mjs";
import ERTS from "../../assets/js/erts.mjs";
import Interpreter from "../../assets/js/interpreter.mjs";
import Serializer from "../../assets/js/serializer.mjs";
import Type from "../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

const deserialize = Deserializer.deserialize;
const serialize = Serializer.serialize;

function testNestedBitstringDeserialization(
  nestedTerm,
  destination = "server",
) {
  const term = {x: nestedTerm};
  const serialized = serialize(term, destination);
  const deserialized = deserialize(serialized);

  assert.equal(typeof deserialized, "object");
  assert.deepStrictEqual(Object.keys(deserialized), ["x"]);
  assert.isTrue(Type.isBitstring(deserialized.x));
  assert.isTrue(Interpreter.isStrictlyEqual(deserialized.x, term.x));
}

function testNestedDeserialization(
  nestedTerm,
  destination = "server",
  expectedNestedTerm = null,
) {
  const term = {x: nestedTerm};
  const serialized = serialize(term, destination);
  const deserialized = deserialize(serialized);

  let expected;

  if (expectedNestedTerm === null) {
    expected = term;
  } else {
    expected = {x: expectedNestedTerm};
  }

  assert.deepStrictEqual(deserialized, expected);
}

function testTopLevelBitstringDeserialization(term, destination = "server") {
  const serialized = serialize(term, destination);
  const deserialized = deserialize(serialized);

  assert.isTrue(Interpreter.isStrictlyEqual(deserialized, term));
}

function testTopLevelDeserialization(
  term,
  destination = "server",
  expected = null,
) {
  const serialized = serialize(term, destination);
  const deserialized = deserialize(serialized);

  if (expected === null) {
    expected = term;
  }

  assert.deepStrictEqual(deserialized, expected);
}

describe("Deserializer", () => {
  describe("deserialize()", () => {
    describe("boxed terms", () => {
      describe("atom", () => {
        const term = Type.atom("abc");

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("bitstring", () => {
        describe("empty", () => {
          const term = Type.bitstring("");

          it("top-level", () => {
            testTopLevelBitstringDeserialization(term);
          });

          it("nested", () => {
            testNestedBitstringDeserialization(term);
          });
        });

        describe("single-byte", () => {
          describe("without leftover bits", () => {
            const term = Type.bitstring("a");

            it("top-level", () => {
              testTopLevelBitstringDeserialization(term);
            });

            it("nested", () => {
              testNestedBitstringDeserialization(term);
            });
          });

          describe("with leftover bits", () => {
            const term = Type.bitstring([1, 0, 1, 0]);

            it("top-level", () => {
              testTopLevelBitstringDeserialization(term);
            });

            it("nested", () => {
              testNestedBitstringDeserialization(term);
            });
          });
        });

        describe("multiple-byte", () => {
          describe("without leftover bits", () => {
            const term = Type.bitstring("Hologram");

            it("top-level", () => {
              testTopLevelBitstringDeserialization(term);
            });

            it("nested", () => {
              testNestedBitstringDeserialization(term);
            });
          });

          describe("with leftover bits", () => {
            const term = Type.bitstring([1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0]);

            it("top-level", () => {
              testTopLevelBitstringDeserialization(term);
            });

            it("nested", () => {
              testNestedBitstringDeserialization(term);
            });
          });
        });
      });

      describe("float", () => {
        describe("encoded as float", () => {
          const term = Type.float(1.23);

          it("top-level", () => {
            testTopLevelDeserialization(term);
          });

          it("nested", () => {
            testNestedDeserialization(term);
          });
        });

        describe("encoded as integer", () => {
          const term = Type.float(123);

          it("top-level", () => {
            testTopLevelDeserialization(term);
          });

          it("nested", () => {
            testNestedDeserialization(term);
          });
        });
      });

      describe("function", () => {
        beforeEach(() => {
          ERTS.funSequence.reset();
        });

        const context = contextFixture({
          module: Type.alias("MyModule"),
          vars: {x: Type.integer(10), y: Type.integer(20)},
        });

        describe("capture", () => {
          const term = Type.functionCapture(
            "Calendar.ISO",
            "date_to_string",
            4,
            [],
            context,
          );

          const expectedContext = contextFixture({
            module: null,
            vars: {},
          });

          const expectedTerm = {...term, context: expectedContext};

          it("top-level", () => {
            testTopLevelDeserialization(term, "server", expectedTerm);
          });

          it("nested", () => {
            testNestedDeserialization(term, "server", expectedTerm);
          });
        });

        describe("non-capture", () => {
          const term = Type.anonymousFunction(
            1,
            [
              {
                params: (_context) => [Type.variablePattern("a")],
                guards: [
                  (context) => Erlang["/=/2"](context.vars.a, Type.integer(1)),
                ],
                // prettier-ignore
                body: (context) => { return Type.list([Type.atom("m"), context.vars.a, context.vars.x]) },
              },
              {
                params: (_context) => [Type.variablePattern("a")],
                guards: [
                  (context) => Erlang["/=/2"](context.vars.a, Type.integer(2)),
                ],
                // prettier-ignore
                body: (context) => { return Type.list([Type.atom("n"), context.vars.a, context.vars.y]) },
              },
            ],
            context,
          );

          it("top-level", () => {
            const serialized = serialize(term, "client");
            const deserialized = deserialize(serialized);

            assert.deepStrictEqual(deserialized, {
              ...term,
              clauses: deserialized.clauses,
            });

            assert.equal(deserialized.clauses.length, 2);

            assert.isFunction(deserialized.clauses[0].params);
            assert.isFunction(deserialized.clauses[0].guards[0]);
            assert.isFunction(deserialized.clauses[0].body);

            assert.isFunction(deserialized.clauses[1].params);
            assert.isFunction(deserialized.clauses[1].guards[0]);
            assert.isFunction(deserialized.clauses[1].body);

            const callResult = Interpreter.callAnonymousFunction(deserialized, [
              Type.integer(1),
            ]);

            const expectedCallResult = Type.list([
              Type.atom("n"),
              Type.integer(1),
              Type.integer(20),
            ]);

            assert.deepStrictEqual(callResult, expectedCallResult);
          });

          it("nested", () => {
            const nestedTerm = {t: term};
            const serialized = serialize(nestedTerm, "client");
            const deserialized = deserialize(serialized);

            assert.deepStrictEqual(deserialized, {
              t: {
                ...term,
                clauses: deserialized.t.clauses,
              },
            });

            assert.equal(deserialized.t.clauses.length, 2);

            assert.isFunction(deserialized.t.clauses[0].params);
            assert.isFunction(deserialized.t.clauses[0].guards[0]);
            assert.isFunction(deserialized.t.clauses[0].body);

            assert.isFunction(deserialized.t.clauses[1].params);
            assert.isFunction(deserialized.t.clauses[1].guards[0]);
            assert.isFunction(deserialized.t.clauses[1].body);

            const callResult = Interpreter.callAnonymousFunction(
              deserialized.t,
              [Type.integer(1)],
            );

            const expectedCallResult = Type.list([
              Type.atom("n"),
              Type.integer(1),
              Type.integer(20),
            ]);

            assert.deepStrictEqual(callResult, expectedCallResult);
          });
        });
      });

      describe("integer", () => {
        const term = Type.integer(90071992547409919007199254740991n);

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("list", () => {
        const term = Type.list([Type.integer(1), Type.float(2.34)]);

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("map", () => {
        const term = Type.map([
          [Type.atom("x"), Type.integer(1)],
          [Type.atom("y"), Type.float(2.34)],
        ]);

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("pid", () => {
        const term = Type.pid('my_node@my_"host', [0, 11, 222], "server");

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("port", () => {
        const term = Type.port('my_node@my_"host', [0, 11], "server");

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("reference", () => {
        const term = Type.reference('my_node@my_"host', 0, [3, 2, 1]);

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("tuple", () => {
        const term = Type.tuple([Type.integer(1), Type.float(2.34)]);

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });
    });

    describe("JS terms", () => {
      describe("array", () => {
        const term = [9, 8.76];

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("float", () => {
        const term = 1.23;

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("function", () => {
        describe("longhand syntax", () => {
          it("top-level", () => {
            const serialized = `[2,"ufunction (a, b) { const result = Type.integer(a + b); return result; }"]`;
            const deserialized = deserialize(serialized);

            assert.isFunction(deserialized);
            assert.deepStrictEqual(deserialized(1, 2), Type.integer(3));
          });

          it("nested", () => {
            const serialized = `[2,{"x":"ufunction (a, b) { const result = Type.integer(a + b); return result; }"}]`;
            const deserialized = deserialize(serialized);

            assert.deepStrictEqual(Object.keys(deserialized), ["x"]);
            assert.isFunction(deserialized.x);
            assert.deepStrictEqual(deserialized.x(1, 2), Type.integer(3));
          });
        });

        describe("shorthand syntax", () => {
          it("top-level", () => {
            const serialized = `[2,"u(a, b) => Type.integer(a + b)"]`;
            const deserialized = deserialize(serialized);

            assert.isFunction(deserialized);
            assert.deepStrictEqual(deserialized(1, 2), Type.integer(3));
          });

          it("nested", () => {
            const serialized = `[2,{"x":"u(a, b) => Type.integer(a + b)"}]`;
            const deserialized = deserialize(serialized);

            assert.deepStrictEqual(Object.keys(deserialized), ["x"]);
            assert.isFunction(deserialized.x);
            assert.deepStrictEqual(deserialized.x(1, 2), Type.integer(3));
          });
        });
      });

      describe("integer", () => {
        const term = 123;

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("null", () => {
        const term = null;

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("object", () => {
        const term = {v: 9, 'x"yz': 8.76};

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });

      describe("string", () => {
        const term = 'a"bc';

        it("top-level", () => {
          testTopLevelDeserialization(term);
        });

        it("nested", () => {
          testNestedDeserialization(term);
        });
      });
    });
  });
});

File: test/elixir/hologram/runtime/deserializer_test.exs
defmodule Hologram.Runtime.DeserializerTest do
  use Hologram.Test.BasicCase, async: true
  import Hologram.Runtime.Deserializer

  @delimiter delimiter()

  describe "version 3" do
    test "top-level data, raw JSON" do
      assert deserialize(~s'[3,"axyz"]') == :xyz
    end

    test "top-level data, already JSON-decoded" do
      assert deserialize([3, "axyz"]) == :xyz
    end

    test "atom" do
      assert deserialize(3, "axyz") == :xyz
    end

    test "bitstring, empty" do
      assert deserialize(3, "b") == ""
    end

    test "bitstring, single-byte, without leftover bits" do
      assert deserialize(3, "b061") == "a"
    end

    test "bitstring, single-byte, with leftover bits" do
      assert deserialize(3, "b3a0") == <<5::size(3)>>
    end

    test "bitstring, multiple-byte, without leftover bits" do
      assert deserialize(3, "b0486f6c6f6772616d") == "Hologram"
    end

    test "bitstring, multiple-byte, with leftover bits" do
      assert deserialize(3, "b3616263a0") == <<97, 98, 99, 5::size(3)>>
    end

    test "float, encoded as float" do
      assert deserialize(3, "f1.23") === 1.23
    end

    test "float, encoded as integer" do
      assert deserialize(3, "f123") === 123.0
    end

    test "function capture" do
      data = "cCalendar.ISO#{@delimiter}parse_date#{@delimiter}2"
      assert deserialize(3, data) == (&Calendar.ISO.parse_date/2)
    end

    test "integer" do
      assert deserialize(3, "i123") == 123
    end

    test "list" do
      data = %{"t" => "l", "d" => ["i1", "f2.34"]}
      assert deserialize(3, data) == [1, 2.34]
    end

    test "map" do
      data = %{"t" => "m", "d" => [["ax", "i1"], ["b079", "f2.34"]]}
      assert deserialize(3, data) == %{:x => 1, "y" => 2.34}
    end

    test "pid" do
      data = "pmy_node@my_host#{@delimiter}0,11,222#{@delimiter}server"
      assert deserialize(3, data) == pid("0.11.222")
    end

    test "port" do
      data = "omy_node@my_host#{@delimiter}0,11#{@delimiter}server"
      assert deserialize(3, data) == port("0.11")
    end

    test "reference" do
      data = %{"t" => "r", "n" => "nonode@nohost", "c" => 0, "i" => [3, 2, 1]}
      assert deserialize(3, data) == ref("0.1.2.3")
    end

    test "tuple" do
      data = %{"t" => "t", "d" => ["i1", "f2.34"]}
      assert deserialize(3, data) == {1, 2.34}
    end
  end

  describe "version 2" do
    test "top-level data, raw JSON" do
      assert deserialize(~s'[2,"axyz"]') == :xyz
    end

    test "top-level data, already JSON-decoded" do
      assert deserialize([2, "axyz"]) == :xyz
    end

    test "atom" do
      assert deserialize(2, "axyz") == :xyz
    end

    test "bitstring, empty" do
      assert deserialize(2, "b") == ""
    end

    test "bitstring, single-byte, without leftover bits" do
      assert deserialize(2, "b061") == "a"
    end

    test "bitstring, single-byte, with leftover bits" do
      assert deserialize(2, "b3a0") == <<5::size(3)>>
    end

    test "bitstring, multiple-byte, without leftover bits" do
      assert deserialize(2, "b0486f6c6f6772616d") == "Hologram"
    end

    test "bitstring, multiple-byte, with leftover bits" do
      assert deserialize(2, "b3616263a0") == <<97, 98, 99, 5::size(3)>>
    end

    test "float, encoded as float" do
      assert deserialize(2, "f1.23") === 1.23
    end

    test "float, encoded as integer" do
      assert deserialize(2, "f123") === 123.0
    end

    test "function capture" do
      data = "cCalendar.ISO#{@delimiter}parse_date#{@delimiter}2"
      assert deserialize(2, data) == (&Calendar.ISO.parse_date/2)
    end

    test "integer" do
      assert deserialize(2, "i123") == 123
    end

    test "list" do
      data = %{"t" => "l", "d" => ["i1", "f2.34"]}
      assert deserialize(2, data) == [1, 2.34]
    end

    test "map" do
      data = %{"t" => "m", "d" => [["ax", "i1"], ["b079", "f2.34"]]}
      assert deserialize(2, data) == %{:x => 1, "y" => 2.34}
    end

    test "pid" do
      data = "pmy_node@my_host#{@delimiter}0,11,222#{@delimiter}server"
      assert deserialize(2, data) == pid("0.11.222")
    end

    test "port" do
      data = "omy_node@my_host#{@delimiter}0,11#{@delimiter}server"
      assert deserialize(2, data) == port("0.11")
    end

    test "reference" do
      data = "rmy_node@my_host#{@delimiter}0,1,2,3#{@delimiter}server"
      assert deserialize(2, data) == ref("0.1.2.3")
    end

    test "tuple" do
      data = %{"t" => "t", "d" => ["i1", "f2.34"]}
      assert deserialize(2, data) == {1, 2.34}
    end
  end

  describe "version 1" do
    test "top-level data, raw JSON" do
      assert deserialize(~s'[1,"__atom__:xyz"]') == :xyz
    end

    test "top-level data, already JSON-decoded" do
      assert deserialize([1, "__atom__:xyz"]) == :xyz
    end

    test "atom" do
      assert deserialize(1, "__atom__:xyz") == :xyz
    end

    test "bitstring, binary" do
      assert deserialize(1, "__binary__:a\"bc") == "a\"bc"
    end

    test "bitstring, non-binary" do
      data = %{"type" => "bitstring", "bits" => [1, 0, 1, 0]}
      assert deserialize(1, data) == <<1::1, 0::1, 1::1, 0::1>>
    end

    test "float, encoded as float" do
      assert deserialize(1, "__float__:1.23") === 1.23
    end

    test "float, encoded as integer" do
      assert deserialize(1, "__float__:123") === 123.0
    end

    test "function capture" do
      data = %{
        "type" => "anonymous_function",
        "capturedModule" => "Calendar.ISO",
        "capturedFunction" => "parse_date",
        "arity" => 2
      }

      assert deserialize(1, data) == (&Calendar.ISO.parse_date/2)
    end

    test "integer" do
      assert deserialize(1, "__integer__:123") == 123
    end

    test "list" do
      data = %{
        "type" => "list",
        "data" => ["__integer__:1", "__float__:2.34"]
      }

      assert deserialize(1, data) == [1, 2.34]
    end

    test "map" do
      data = %{
        "type" => "map",
        "data" => [
          ["__atom__:a", "__integer__:1"],
          ["__binary__:b", "__float__:2.34"]
        ]
      }

      assert deserialize(1, data) == %{:a => 1, "b" => 2.34}
    end

    test "pid" do
      data = %{"type" => "pid", "segments" => [0, 11, 222]}
      assert deserialize(1, data) == pid("0.11.222")
    end

    test "port" do
      data = %{"type" => "port", "value" => "0.11"}
      assert deserialize(1, data) == port("0.11")
    end

    test "reference" do
      data = %{"type" => "reference", "value" => "0.1.2.3"}
      assert deserialize(1, data) == ref("0.1.2.3")
    end

    test "tuple" do
      data = %{
        "type" => "tuple",
        "data" => ["__integer__:1", "__float__:2.34"]
      }

      assert deserialize(1, data) == {1, 2.34}
    end
  end
end
